source/server/hot_restart_impl.h

```diff
@@ -136,76 +140,11 @@ class HotRestartImpl : public HotRestart, Logger::Loggable<Logger::Id::main> {
   static std::string hotRestartVersion(uint64_t max_num_stats, uint64_t max_stat_name_len);
 
 private:
-  enum class RpcMessageType {
-    DrainListenersRequest = 1,
-    GetListenSocketRequest = 2,
-    GetListenSocketReply = 3,
-    ShutdownAdminRequest = 4,
-    ShutdownAdminReply = 5,
-    TerminateRequest = 6,
-    UnknownRequestReply = 7,
-    GetStatsRequest = 8,
-    GetStatsReply = 9
-  };
-
-  PACKED_STRUCT(struct RpcBase {
-    RpcBase(RpcMessageType type, uint64_t length = sizeof(RpcBase))
-        : type_(type), length_(length) {}
-
-    RpcMessageType type_;
-    uint64_t length_;
-  });
-
-  PACKED_STRUCT(struct RpcGetListenSocketRequest
-                : public RpcBase {
-                  RpcGetListenSocketRequest()
-                      : RpcBase(RpcMessageType::GetListenSocketRequest, sizeof(*this)) {}
-
-                  char address_[256]{0};
-                });
-
-  PACKED_STRUCT(struct RpcGetListenSocketReply
-                : public RpcBase {
-                  RpcGetListenSocketReply()
-                      : RpcBase(RpcMessageType::GetListenSocketReply, sizeof(*this)) {}
-
-                  int fd_{0};
-                });
-
-  PACKED_STRUCT(struct RpcShutdownAdminReply
-                : public RpcBase {
-                  RpcShutdownAdminReply()
-                      : RpcBase(RpcMessageType::ShutdownAdminReply, sizeof(*this)) {}
-
-                  uint64_t original_start_time_{0};
-                });
-
-  PACKED_STRUCT(struct RpcGetStatsReply
-                : public RpcBase {
-                  RpcGetStatsReply() : RpcBase(RpcMessageType::GetStatsReply, sizeof(*this)) {}
-
-                  uint64_t memory_allocated_{0};
-                  uint64_t num_connections_{0};
-                  uint64_t unused_[16]{0};
-                });
-
-  template <class rpc_class, RpcMessageType rpc_type> rpc_class* receiveTypedRpc() {
-    RpcBase* base_message = receiveRpc(true);
-    RELEASE_ASSERT(base_message->length_ == sizeof(rpc_class), "");
-    RELEASE_ASSERT(base_message->type_ == rpc_type, "");
-    return reinterpret_cast<rpc_class*>(base_message);
-  }
-
-  int bindDomainSocket(uint64_t id);
-  void initDomainSocketAddress(sockaddr_un* address);
-  sockaddr_un createDomainSocketAddress(uint64_t id);
-  void onGetListenSocket(RpcGetListenSocketRequest& rpc);
-  void onSocketEvent();
-  RpcBase* receiveRpc(bool block);
-  void sendMessage(sockaddr_un& address, RpcBase& rpc);
   static std::string versionHelper(uint64_t max_num_stats, const Stats::StatsOptions& stats_options,
                                    Stats::RawStatDataSet& stats_set);
 
+  HotRestartingChild as_child_;
+  HotRestartingParent as_parent_;
```

Comment:
The two "views" that I have split most of the hot restart logic into: the logic relevant when we are acting as child to an existing parent, and the logic relevant when we are parent to a new child.